dataclients/kubernetes/traffic.go

```diff
@@ -0,0 +1,122 @@
+package kubernetes
+
+import (
+	"github.com/zalando/skipper/dataclients/kubernetes/definitions"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/predicates"
+)
+
+// backendTraffic specifies whether a given backend is allowed to receive any traffic and
+// modifies route to receive the desired traffic portion
+type backendTraffic interface {
+	allowed() bool
+	apply(*eskip.Route)
+}
+
+// GetBackendTrafficCalculator returns a function that calculates backendTraffic for each backend using specified algorithm
+func GetBackendTrafficCalculator[T definitions.WeightedBackend](algorithm string) func(b []T) map[string]backendTraffic {
+	// WIP: only traffic predicate calculator is supported for now
+	return trafficPredicateCalculator[T]
+}
+
+// trafficPredicate implements backendTraffic using Traffic() and True() predicates
+type trafficPredicate struct {
+	value   float64
+	balance int
+}
+
+var _ backendTraffic = &trafficPredicate{}
+
+// trafficPredicateCalculator calculates argument for the Traffic() predicate and
+// the number of True() predicates to be added to the routes based on the weights of the backends.
+//
+// The Traffic() argument is calculated based on the following rules:
+//
+//   - if no weight is defined for a backend it will get weight 0.
+//   - if no weights are specified for all backends of a path, then traffic will
+//     be distributed equally.
+//
+// Each Traffic() argument is relative to the number of remaining backends,
+// e.g. if the weight is specified as:
+//
+//	backend-1: 0.1
+//	backend-2: 0.2
+//	backend-3: 0.3
+//	backend-4: 0.4
+//
+// then Traffic() predicate arguments will be:
+//
+//	backend-1: Traffic(0.1)   == 0.1 / (0.1 + 0.2 + 0.3 + 0.4)
+//	backend-2: Traffic(0.222) == 0.2 / (0.2 + 0.3 + 0.4)
+//	backend-3: Traffic(0.428) == 0.3 / (0.3 + 0.4)
+//	backend-4: Traffic(1.0)   == 0.4 / (0.4)
+//
+// The weight of the backend routes will be adjusted by a number of True() predicates
+// equal to the number of remaining backends minus one, e.g. for the above example:
+//
+//	backend-1: Traffic(0.1)   && True() && True() -> ...
+//	backend-2: Traffic(0.222) && True() -> ...
+//	backend-3: Traffic(0.428) -> ...
+//	backend-4: Traffic(1.0)   -> ...
+//
+// Traffic(1.0) is handled in a special way, see trafficPredicate.apply().
+func trafficPredicateCalculator[T definitions.WeightedBackend](b []T) map[string]backendTraffic {
+	sum := 0.0
+	weights := make([]float64, len(b))
+	for i, bi := range b {
+		w := bi.GetWeight()
+		weights[i] = w
+		sum += w
+	}
+
+	if sum == 0 {
+		sum = float64(len(weights))
+		for i := range weights {
+			weights[i] = 1
+		}
+	}
+
+	var lastWithWeight int
+	for i, w := range weights {
+		if w > 0 {
+			lastWithWeight = i
+		}
+	}
+
+	bt := make(map[string]backendTraffic)
+
+	for i, bi := range b {
```

Comment:
I can't map `bi` name to anything

Reply:
Probably it means "Backend Index". it is used just 2 lines below:
```
bt[bi.GetName()] = ct
```

Reply:
bi == item of b